### ISSUE TITLE
Stack traces not properly showing up for Flutter web

### ISSUE THREAD

**Issue by venkatd** [getsentry/sentry-dart#562]
_Platform:_
- [ ]  Dart
- [ ]  Flutter Android or iOS
- [x]  Flutter Web

_IDE:_
- [ ]  VSCode
- [x]  IntelliJ/AS
- [ ]  XCode
- [ ]  Other, which one?

_split-debug-info and obfuscate (Flutter Android or iOS) or CanvasKit (Flutter Web):_
- [ ]  Enabled
- [ ]  Disabled

_Platform installed with:_
- [x] pub.dev
- [ ] GitHub

Output of the command `flutter doctor -v` below:

<details>
<summary>flutter doctor -v</summary>

```
$ flutter doctor -v
[✓] Flutter (Channel unknown, 2.4.0-5.0.pre.161, on Mac OS X 10.14.6 18G9323
    darwin-x64, locale en-US)
    • Flutter version 2.4.0-5.0.pre.161 at /Users/venkat/code/flutter
    • Upstream repository unknown
    • Framework revision b81165d259 (4 weeks ago), 2021-07-22 14:55:18 -0700
    • Engine revision 1402e43871
    • Dart version 2.14.0 (build 2.14.0-343.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for
      more details.

[!] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C504
    ✗ Xcode 11.3.1 out of date (12.0.1 is recommended).
      Download the latest version or update via the Mac App Store.
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.4)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 38.2.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build
      1.8.0_152-release-1343-b01)

[✓] IntelliJ IDEA Community Edition (version 2020.1.4)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 55.0.1
    • Dart plugin version 201.9335

[✓] VS Code (version 1.59.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • Mac OS X 10.14.6 18G9323
      darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 92.0.4515.159

! Doctor found issues in 2 categories.
```
</details>

The version of the SDK (See pubspec.lock):
**6.0.0-beta.3**

---
I have the following issue:

Stack traces are not properly showing up do to an error with the source maps.

**Steps to reproduce:**

We have a GitHub action setup as so:
<details>
<summary>deploy-web.yml</summary>

```
name: Deploy Turtle web app to Netlify

on:
  push:
    branches:
      - master

jobs:
  build:
    runs-on: ubuntu-latest
    defaults:
      run:
        working-directory: packages/turtle_app
    steps:
    - uses: actions/checkout@v2
    - uses: TurtleAI/flutter-action@allow-specifying-master-commit
      with:
        channel: master
        flutter-version: b81165d259b4428af4efa47183842b577bd389c1
        platforms: web
    - name: Cache dependencies
      uses: actions/cache@v2
      id: deps-cache
      with:
        path: ${{ env.PUB_CACHE }}
        key: ${{ runner.os }}-pub-${{ hashFiles('**/pubspec.lock') }}
        restore-keys: ${{ runner.os }}-pub-
    - name: Fetch dependencies
      run: flutter pub get
    - name: Build app
      run:  flutter build web --verbose --profile --source-maps --dart-define=FLUTTER_APP_COMMIT=$GITHUB_SHA --dart-define=SENTRY_RELEASE=$GITHUB_SHA
    - name: Fingerprint main.dart.js
      run: |
        mv ./build/web/main.dart.js ./build/web/main-${GITHUB_SHA}.dart.js
        mv ./build/web/main.dart.js.map ./build/web/main-${GITHUB_SHA}.dart.js.map
        sed -i -e "s/main\.dart\.js/main-${GITHUB_SHA}.dart.js/g" ./build/web/index.html
        sed -i -e "s/sourceMappingURL=main\.dart\.js\.map/sourceMappingURL=main-${GITHUB_SHA}.dart.js.map/g" ./build/web/main-${GITHUB_SHA}.dart.js
    - name: Deploy to Netlify
      uses: nwtgck/actions-netlify@v1.1
      with:
        publish-dir: ${{ github.workspace }}/packages/turtle_app/build/web
        production-branch: master
        github-token: ${{ secrets.MY_GITHUB_SECRET }}
        deploy-message: ${{ github.event.commits[0].message }}
        netlify-config-path: ${{ github.workspace }}/packages/turtle_app/netlify.toml
        enable-pull-request-comment: false
        enable-commit-comment: false
      env:
        NETLIFY_AUTH_TOKEN: ${{ secrets.NETLIFY_AUTH_TOKEN }}
        NETLIFY_SITE_ID: ${{ secrets.NETLIFY_SITE_ID }}
      timeout-minutes: 1
    - name: Sentry Release
      uses: getsentry/action-release@v1.1.5
      env:
        SENTRY_AUTH_TOKEN: ${{ secrets.SENTRY_AUTH_TOKEN }}
        SENTRY_ORG: ${{ secrets.SENTRY_ORG }}
        SENTRY_PROJECT: ${{ secrets.SENTRY_PROJECT }}
      with:
        environment: prod
        version: ${{ github.sha }}
        sourcemaps: ${{ github.workspace }}/packages/turtle_app/build/web
```
</details>

The sentry release part of our script looks like so:

<details>
<summary>Sentry release logs</summary>

```
##[debug]Evaluating: secrets.SENTRY_AUTH_TOKEN
##[debug]Evaluating Index:
##[debug]..Evaluating secrets:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'SENTRY_AUTH_TOKEN'
##[debug]=> '***'
##[debug]Result: '***'
##[debug]Evaluating: secrets.SENTRY_ORG
##[debug]Evaluating Index:
##[debug]..Evaluating secrets:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'SENTRY_ORG'
##[debug]=> '***'
##[debug]Result: '***'
##[debug]Evaluating: secrets.SENTRY_PROJECT
##[debug]Evaluating Index:
##[debug]..Evaluating secrets:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'SENTRY_PROJECT'
##[debug]=> '***'
##[debug]Result: '***'
##[debug]Evaluating condition for step: 'Sentry Release'
##[debug]Evaluating: success()
##[debug]Evaluating success:
##[debug]=> true
##[debug]Result: true
##[debug]Starting: Sentry Release
##[debug]Loading inputs
##[debug]Evaluating: github.sha
##[debug]Evaluating Index:
##[debug]..Evaluating github:
##[debug]..=> Object
##[debug]..Evaluating String:
##[debug]..=> 'sha'
##[debug]=> 'fce4796231e6aced5cf5d8eca27d589d49fa58cc'
##[debug]Result: 'fce4796231e6aced5cf5d8eca27d589d49fa58cc'
##[debug]Evaluating: format('{0}/packages/***_app/build/web', github.workspace)
##[debug]Evaluating format:
##[debug]..Evaluating String:
##[debug]..=> '{0}/packages/***_app/build/web'
##[debug]..Evaluating Index:
##[debug]....Evaluating github:
##[debug]....=> Object
##[debug]....Evaluating String:
##[debug]....=> 'workspace'
##[debug]..=> '/home/runner/work/***-frontend/***-frontend'
##[debug]=> '/home/runner/work/***-frontend/***-frontend/packages/***_app/build/web'
##[debug]Result: '/home/runner/work/***-frontend/***-frontend/packages/***_app/build/web'
##[debug]Loading env
Run getsentry/action-release@v1.1.5
/usr/bin/docker run --name sentryintegrationssentrygithubactionreleaselatest_8ac21e --label c9e036 --workdir /github/workspace --rm -e FLUTTER_ROOT -e PUB_CACHE -e SENTRY_AUTH_TOKEN -e SENTRY_ORG -e SENTRY_PROJECT -e INPUT_ENVIRONMENT -e INPUT_VERSION -e INPUT_SOURCEMAPS -e INPUT_FINALIZE -e INPUT_STARTED_AT -e INPUT_VERSION_PREFIX -e INPUT_SET_COMMITS -e INPUT_PROJECTS -e INPUT_URL_PREFIX -e HOME -e GITHUB_JOB -e GITHUB_REF -e GITHUB_SHA -e GITHUB_REPOSITORY -e GITHUB_REPOSITORY_OWNER -e GITHUB_RUN_ID -e GITHUB_RUN_NUMBER -e GITHUB_RETENTION_DAYS -e GITHUB_ACTOR -e GITHUB_WORKFLOW -e GITHUB_HEAD_REF -e GITHUB_BASE_REF -e GITHUB_EVENT_NAME -e GITHUB_SERVER_URL -e GITHUB_API_URL -e GITHUB_GRAPHQL_URL -e GITHUB_WORKSPACE -e GITHUB_ACTION -e GITHUB_EVENT_PATH -e GITHUB_ACTION_REPOSITORY -e GITHUB_ACTION_REF -e GITHUB_PATH -e GITHUB_ENV -e RUNNER_DEBUG -e RUNNER_OS -e RUNNER_TOOL_CACHE -e RUNNER_TEMP -e RUNNER_WORKSPACE -e ACTIONS_RUNTIME_URL -e ACTIONS_RUNTIME_TOKEN -e ACTIONS_CACHE_URL -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/***-frontend/***-frontend":"/github/workspace" sentryintegrations/sentry-github-action-release:latest
##[debug]Version is fce4796231e6aced5cf5d8eca27d589d49fa58cc
##[debug]Setting commits with option 'auto'
##[debug]Adding sourcemaps
> Found 5 release files
> Analyzing 5 sources
> Rewriting sources
> Adding source map references
> Bundled 5 files for upload
> Uploaded release files to Sentry
> File upload complete (processing pending on server)

Source Map Upload Report
  Scripts
    /flutter_service_worker.js
    /main-fce4796231e6aced5cf5d8eca27d589d49fa58cc.dart.js
    /post-message.js
    /service-worker.js
  Source Maps
    /main-fce4796231e6aced5cf5d8eca27d589d49fa58cc.dart.js.map
##[debug]Adding deploy to release
##[debug]Finalizing the release
##[debug]Done

::set-output name=version::fce4796231e6aced5cf5d8eca27d589d49fa58cc
##[debug]='fce4796231e6aced5cf5d8eca27d589d49fa58cc'
##[debug]Docker Action run completed with exit code 0
##[debug]Finishing: Sentry Release
```
</details>

**Actual result:**
![130146638-b8c0ea53-efad-43c1-9c55-9108061f19e6](https://user-images.githubusercontent.com/220686/130274105-710b86aa-ece1-4470-8d14-a793d76fa31b.png)

If we trigger an exception we don't get any stack traces.

**Expected result:**
I should be able to see the full source code for each line of the stack trace.

## I also tried setting this up from the command line and it gave me errors

```
flutter build web --verbose --profile --source-maps --dart-define=FLUTTER_APP_COMMIT=$GITHUB_SHA --dart-define=SENTRY_RELEASE=$GITHUB_SHA
```

Output:
https://gist.github.com/venkatd/ff0094f602235b0b5ad03557782fa383

Then the command to upload the dart files:

```
sentry-cli releases files $SENTRY_RELEASE upload-sourcemaps . --ext dart --rewrite --validate --log-level=trace
```

Output:
https://gist.github.com/venkatd/3234a96f02d2e12035adfed2dfe57496

Then the command to upload the javascript files (ends up with warnings on missing sources):

```
sentry-cli releases files $SENTRY_RELEASE upload-sourcemaps . --ext map --ext js --rewrite --validate --log-level=trace
```

Output:
https://gist.github.com/venkatd/fcc6f61d74b30ef39c731d56701e7f57

**Comment by ueman** [comment]
Hey thanks for filing the issue. I see that you're using the master channel. Does it work if you're using the stable channel?

**Comment by venkatd** [comment]
@ueman 

We're on the master channel to take advantage of the recent improvements such as collecting browser information.

If I check old sentry events from a few days ago on `5.1.0` we weren't able to see stack traces either.

Here are screenshots of it:

![image](https://user-images.githubusercontent.com/220686/130296550-2dd948b3-9e2c-4443-8ce6-9be72eedb012.png)

![image](https://user-images.githubusercontent.com/220686/130296616-c20c657a-c9ff-4722-b9a9-4cf4faaba8f3.png)


![image](https://user-images.githubusercontent.com/220686/130296570-100ab3ee-e70f-4031-a479-adee6f2ebb17.png)

**Comment by ueman** [comment]
I wanted to know if it works when you're on Flutters stable channel. Sorry for the confusion.

**Comment by venkatd** [comment]
@ueman no problem!

Would there be a recommended way to test this without messing with our production setup? (We wouldn't want to update to the stable channel on our production release.)

**Comment by marandaneto** [comment]
@venkatd the source map, event fields, and release should match in order to work it properly.

the event.release and event.dist (set by the SDK or dart define) should match the created release by sentrry-cli.
also, renaming the files by appending a GITHUB_SHA suffix could be another problem.

could you share an event on sentry.io so we can verify some other things? thanks

**Comment by venkatd** [comment]
Sure, here's an event:

https://sentry.io/share/issue/c50c8b324e974332ad857fd5b031b2a2/

```
  "release": "4051bbb25374c90edc337acacc5221db09736ce0",
  "dist": "2000023",
```

What is the role of release and dist? How can I confirm whether they match?

Perhaps I'm misunderstanding their use. I have been setting the release to the commit hash so we can know the exact commit each exception came from for easier reproduction.

For us:
release: currently the github commit sha as in `git rev-parse head`
dist: comes from our pubspec.yaml (shown below) file
```yaml
name: turtle_app
description: Turtle application

version: 3.5.1+2000023
```

Is there a recommended way to setup releases for flutter web? I think it may be best to drop down to using the sentry cli directly because the the GitHub action doesn't allow specifying extensions.

I'm looking at the example script here:
https://github.com/getsentry/sentry-dart/blob/main/flutter/example/run.sh

**Comment by marandaneto** [comment]
indeed, https://github.com/getsentry/sentry-dart/blob/main/flutter/example/run.sh isn't up to date, but check out https://github.com/getsentry/sentry-dart-plugin/blob/main/lib/sentry_dart_plugin.dart#L159-L163

you have to pass `--dist 2000023` then to `sentry-cli` when calling the `releases` command.
Similar issue to https://github.com/getsentry/sentry-dart/issues/512#issuecomment-872195104

you got the meaning of `release` correctly but the `event.release` and creating the release via sentry-cli should be the same to make source maps work.

so if your `event.release` is `4051bbb25374c90edc337acacc5221db09736ce0`, then `SENTRY_RELEASE` should have the same value, or the other around, set the SDK to match the value of `SENTRY_RELEASE`.

also as I said before, do not rename the `*js` and `*js.map` files, they should match the path in the event, so `main` != `main-4051bbb25374c90edc337acacc5221db09736ce0`, it would fail because it does not find the file.

try out and let us know, thanks

**Comment by venkatd** [comment]
Thanks for the details. I'll give this a try.

Do you know any workarounds if we have to rename the `*.js` and `*.js.map` files? We are doing a "poor mans fingerprinting" with to force cache busting.

Perhaps there is a way to avoid renaming things on the sentry side but on our CDN, serve the file with the fingerprinted name? How does sentry typically handle fingerprinted assets?

Here's the unresolved issue that made us resort to renaming the files: https://github.com/flutter/flutter/issues/64968

**Comment by venkatd** [comment]
I've made a bit of progress but there are a few more minor issues remaining.

Please take a look at this more recent event when you get a chance:
https://sentry.io/share/issue/c2c17fb884e945489a2383e47a056724/

The command I used to generate the build is here:
```
sentry-cli releases new $GITHUB_SHA
sentry-cli releases files $GITHUB_SHA upload-sourcemaps . --ext dart --rewrite
sentry-cli releases files $GITHUB_SHA upload-sourcemaps . --ext map --ext js --rewrite
sentry-cli releases finalize $GITHUB_SHA
```

I got rid of the Sentry Release GitHub action and dropped down to `sentry-cli` so I could more closely match what `sentry_dart_plugin` was doing.

The stack trace for our own code seems to work well:
![image](https://user-images.githubusercontent.com/220686/130337238-762cb4c2-8372-47a4-8931-def2ce3b50a7.png)

However, you'll see that the Flutter framework itself isn't being picked up. I get errors like this:

![image](https://user-images.githubusercontent.com/220686/130337300-11eac0dd-0357-4f67-84ab-f51a456c42ad.png)

One thing to note is that somehow `/opt/hostedtoolcache/flutter/...` has snuck into the path. This is the path where the Flutter framework is cached on GitHub actions.

What I have *not done yet*:
* I haven't configured the dist property yet
* The file is still being renamed. It would be good to understand the implications of this. I should mention I have been also renaming the `//# sourceMappingURL` attribute

Any thoughts on next steps? When I get a chance I will check that dist match, but I'm not sure about the impact of having the file renamed.

**Comment by marandaneto** [comment]
> Do you know any workarounds if we have to rename the `*.js` and `*.js.map` files? We are doing a "poor mans fingerprinting" with to force cache busting.

sentry-cli is able to do that as well, but it requires a new conf.
https://docs.sentry.io/product/cli/releases/#upload-files

look at this example:

> The first argument to upload is the path to the file, the second is an optional URL we should associate it with. Note that if you want to use an abbreviated URL (eg: ~/foo.js) make sure to use single quotes to avoid the expansion by the shell to your home folder.

**Comment by marandaneto** [comment]
> I've made a bit of progress but there are a few more minor issues remaining.
> 
> Please take a look at this more recent event when you get a chance:
> https://sentry.io/share/issue/c2c17fb884e945489a2383e47a056724/
> 
> The command I used to generate the build is here:
> 
> ```
> sentry-cli releases new $GITHUB_SHA
> sentry-cli releases files $GITHUB_SHA upload-sourcemaps . --ext dart --rewrite
> sentry-cli releases files $GITHUB_SHA upload-sourcemaps . --ext map --ext js --rewrite
> sentry-cli releases finalize $GITHUB_SHA
> ```

you can ignore `--rewrite` since its the default behavior already

> I got rid of the Sentry Release GitHub action and dropped down to `sentry-cli` so I could more closely match what `sentry_dart_plugin` was doing.
> 
> The stack trace for our own code seems to work well:
> ![image](https://user-images.githubusercontent.com/220686/130337238-762cb4c2-8372-47a4-8931-def2ce3b50a7.png)
> 
> However, you'll see that the Flutter framework itself isn't being picked up. I get errors like this:
> 
> ![image](https://user-images.githubusercontent.com/220686/130337300-11eac0dd-0357-4f67-84ab-f51a456c42ad.png)

the Dart and Flutter source code isn't available, you can either upload the source code yourself (they should match the reported path), or just don't, ideally, those frames would be hidden since its not that important, the error is most likely in your code and not in the Dart/Flutter themselves.

> One thing to note is that somehow `/opt/hostedtoolcache/flutter/...` has snuck into the path. This is the path where the Flutter framework is cached on GitHub actions.
> 
> What I have _not done yet_:
> 
> * I haven't configured the dist property yet
> * The file is still being renamed. It would be good to understand the implications of this. I should mention I have been also renaming the `//# sourceMappingURL` attribute

no `dist` set when creating releases or renaming the files may generate problems locating the correct source maps, the side effect is bad symbolication or no symbolication at all, I believe it'd be more problematic if you have more than one `dist` per `release`, otherwise, if you test it and it works without dist or avoiding the renaming, it should be fine, although I strongly recommend avoiding further issues.

> Any thoughts on next steps? When I get a chance I will check that dist match, but I'm not sure about the impact of having the file renamed.

**Comment by marandaneto** [comment]
looks like the issue is fixed and the comments above are the expected behavior since you didn't upload the Dart and Flutter source code, feel free to close the issue if that helped, thanks

**Comment by venkatd** [comment]
Will do, thanks a lot for the responsiveness and help :)

**Comment by venkatd** [comment]
I'll make sure dist matches up.

Would I have to upload the Flutter source each time or only when it changes? If this pretty involved, perhaps I could create a separate issue once we want to enhance the stack traces.

**Comment by venkatd** [comment]
@marandaneto Actually is it possible that I am mixing up release and dist? It looks to me like here:

https://github.com/getsentry/sentry-react-native/issues/516#issuecomment-462393231

> 
> release = Name of the release (full name, you have to check to url in Sentry when clicking on a release)
> 
> dist = fragmenting a release, you can have one release with multiple dist (Distributions) in there
> Used for appcenter and stuff where you have release: 1.0 dist: 1 (build number for example)
> 

So it feels more close to be using the version number for the release (like 3.5.1) and the commit would be the dist (like build number). And a single release can have multiple dist values.

Can you please confirm?

**Comment by marandaneto** [comment]
> I'll make sure dist matches up.
> 
> Would I have to upload the Flutter source each time or only when it changes? If this pretty involved, perhaps I could create a separate issue once we want to enhance the stack traces.

I guess you have to (at least for now), because of source maps and minification, not entirely sure if minification also affects the Dart lang and the Flutter framework

**Comment by marandaneto** [comment]
> @marandaneto Actually is it possible that I am mixing up release and dist? It looks to me like here:
> 
> [getsentry/sentry-react-native#516 (comment)](https://github.com/getsentry/sentry-react-native/issues/516#issuecomment-462393231)
> 
> > release = Name of the release (full name, you have to check to url in Sentry when clicking on a release)
> > dist = fragmenting a release, you can have one release with multiple dist (Distributions) in there
> > Used for appcenter and stuff where you have release: 1.0 dist: 1 (build number for example)
> 
> So it feels more close to be using the version number for the release (like 3.5.1) and the commit would be the dist (like build number). And a single release can have multiple dist values.
> 
> Can you please confirm?

yes, the release follows the format `my_app@version` or `my_app@version+buildNumber` and the `buildNumber` is also the `dist`, eg `io.sentry.app@1.0.0+854`
in that case: `release=io.sentry.app@1.0.0+854` and `dist=854`, so a release may contain multiple `dist`
